hls-plugin-api/src/Ide/Types.hs

```diff
@@ -180,18 +181,74 @@ instance PluginMethod TextDocumentRangeFormatting where
+
+-- | Notifications that can be handled by plugins.
+-- 'ExtraParams' captures any extra data the IDE passes to the handlers for this method
+class HasTracing (MessageParams m) => PluginNotification m where
+
+  -- | Parse the configuration to check if this plugin is enabled
+  pluginEnabledNotification :: SMethod m -> PluginId -> Config -> Bool
+
+instance PluginNotification TextDocumentDidOpen where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification TextDocumentDidChange where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification TextDocumentDidSave where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification TextDocumentDidClose where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification WorkspaceDidChangeWatchedFiles where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification WorkspaceDidChangeWorkspaceFolders where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification WorkspaceDidChangeConfiguration where
+    pluginEnabledNotification _ _ _ = True
+
+instance PluginNotification Initialized where
+    pluginEnabledNotification _ _ _ = True
+
+-- ---------------------------------------------------------------------
+
 -- | Methods which have a PluginMethod instance
 data IdeMethod (m :: Method FromClient Request) = PluginMethod m => IdeMethod (SMethod m)
 instance GEq IdeMethod where
   geq (IdeMethod a) (IdeMethod b) = geq a b
 instance GCompare IdeMethod where
   gcompare (IdeMethod a) (IdeMethod b) = gcompare a b
 
+-- | Methods which have a PluginMethod instance
+data IdeNotification (m :: Method FromClient Notification) = PluginNotification m => IdeNotification (SMethod m)
+instance GEq IdeNotification where
+  geq (IdeNotification a) (IdeNotification b) = geq a b
+instance GCompare IdeNotification where
+  gcompare (IdeNotification a) (IdeNotification b) = gcompare a b
+
 -- | Combine handlers for the
 newtype PluginHandler a (m :: Method FromClient Request)
   = PluginHandler (PluginId -> a -> MessageParams m -> LspM Config (NonEmpty (Either ResponseError (ResponseResult m))))
 
-newtype PluginHandlers a = PluginHandlers (DMap IdeMethod (PluginHandler a))
+newtype PluginNotificationHandler a (m :: Method FromClient Notification)
+  = PluginNotificationHandler (PluginId -> a -> MessageParams m -> LspM Config (NonEmpty ()))
+-- newtype PluginNotificationHandler a (m :: Method FromClient Notification)
+--   = PluginNotificationHandler (PluginNotificationMethodHandler a m)`
+
+{-
+From Zubin
+alanz_: I would say `newtype PluginNotificationHandler a (m :: Method FromClient Notification) = PluginNotificationHandler (PluginNotificationMethodHandler a m)`
+16:28 and `newtype PluginNotificationHandlers a = PluginNotificationHandlers (DMap SMethod (PluginNotificationHandler a))
+
+-}
+
+newtype PluginHandlers a             = PluginHandlers             (DMap IdeMethod       (PluginHandler a))
```

Comment:
This section from here down is a repeat of what is in  ghcide/src/Development/IDE/Plugin/HLS.hs

Reply:
Which one is redundant?

Reply:
Just to clarify, I haven't added anything. If there is any duplication, it was already present in your branch.

Reply:
fair enough :)